ui/dashboards/src/utils/panelUtils.ts

```diff
@@ -0,0 +1,96 @@
+// Copyright 2023 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { PanelDefinition } from '@perses-dev/core';
+import { PanelGroupDefinition } from '../context';
+
+export const removeWhiteSpacesAndSpecialCharacters = (str: string) => {
+  return str.replace(/\s+/g, '');
+};
+
+// Given a PanelGroup, will find the Y coordinate for adding a new row to the grid, taking into account the items present
+export function getYForNewRow(group: PanelGroupDefinition) {
+  let newRowY = 0;
+  for (const layout of group.itemLayouts) {
+    const itemMaxY = layout.y + layout.h;
+    if (itemMaxY > newRowY) {
+      newRowY = itemMaxY;
+    }
+  }
+  return newRowY;
+}
+
```

Comment:
Helpers below extracted from `panel-editor-slice` and significantly modified to address the edge case I called out in the tests.